libfwupd/fwupd-bios-attr.c

```diff
@@ -0,0 +1,878 @@
+
+	g_return_val_if_fail(FWUPD_IS_BIOS_ATTR(self), NULL);
+
+	g_variant_builder_init(&builder, G_VARIANT_TYPE_VARDICT);
+	if (priv->type != FWUPD_BIOS_ATTR_TYPE_UNKNOWN) {
+		g_variant_builder_add(&builder,
+				      "{sv}",
+				      FWUPD_RESULT_KEY_BIOS_ATTR_TYPE,
+				      g_variant_new_uint32(priv->type));
+	}
+	if (priv->name != NULL) {
+		g_variant_builder_add(&builder,
+				      "{sv}",
+				      FWUPD_RESULT_KEY_NAME,
+				      g_variant_new_string(priv->name));
+	}
+	if (priv->path != NULL) {
+		g_variant_builder_add(&builder,
+				      "{sv}",
+				      FWUPD_RESULT_KEY_FILENAME,
+				      g_variant_new_string(priv->path));
+	}
+	if (priv->description != NULL) {
+		g_variant_builder_add(&builder,
+				      "{sv}",
+				      FWUPD_RESULT_KEY_DESCRIPTION,
+				      g_variant_new_string(priv->description));
+	}
+	if (priv->possible_values->len > 0) {
+		g_autofree const gchar **strv =
+		    g_new0(const gchar *, priv->possible_values->len + 1);
+		for (guint i = 0; i < priv->possible_values->len; i++)
+			strv[i] = (const gchar *)g_ptr_array_index(priv->possible_values, i);
+		g_variant_builder_add(&builder,
+				      "{sv}",
+				      FWUPD_RESULT_KEY_BIOS_ATTR_POSSIBLE_VALUES,
+				      g_variant_new_strv(strv, -1));
+	}
+	return g_variant_new("a{sv}", &builder);
+}
+
+static void
+fwupd_bios_attr_from_key_value(FwupdBiosAttr *self, const gchar *key, GVariant *value)
+{
+	if (g_strcmp0(key, FWUPD_RESULT_KEY_BIOS_ATTR_TYPE) == 0) {
+		fwupd_bios_attr_set_kernel_type(self, g_variant_get_uint64(value));
+		return;
+	}
+	if (g_strcmp0(key, FWUPD_RESULT_KEY_NAME) == 0) {
+		fwupd_bios_attr_set_name(self, g_variant_get_string(value, NULL));
+		return;
+	}
+	if (g_strcmp0(key, FWUPD_RESULT_KEY_DESCRIPTION) == 0) {
+		fwupd_bios_attr_set_description(self, g_variant_get_string(value, NULL));
+		return;
+	}
+	if (g_strcmp0(key, FWUPD_RESULT_KEY_BIOS_ATTR_POSSIBLE_VALUES) == 0) {
+		g_autofree const gchar **strv = g_variant_get_strv(value, NULL);
+		for (guint i = 0; strv[i] != NULL; i++)
+			fwupd_bios_attr_add_possible_value(self, strv[i]);
+		return;
+	}
+	if (g_strcmp0(key, FWUPD_RESULT_KEY_BIOS_ATTR_LOWER_BOUND) == 0) {
+		fwupd_bios_attr_set_lower_bound(self, g_variant_get_uint64(value));
+		return;
+	}
+	if (g_strcmp0(key, FWUPD_RESULT_KEY_BIOS_ATTR_UPPER_BOUND) == 0) {
+		fwupd_bios_attr_set_upper_bound(self, g_variant_get_uint64(value));
+		return;
+	}
+	if (g_strcmp0(key, FWUPD_RESULT_KEY_BIOS_ATTR_SCALAR_INCREMENT) == 0) {
+		fwupd_bios_attr_set_scalar_increment(self, g_variant_get_uint64(value));
+		return;
+	}
+}
+
+/* TODO: move both of these out somewhere more common */
+static void
+fwupd_pad_kv_str(GString *str, const gchar *key, const gchar *value)
+{
+	/* ignore */
+	if (key == NULL || value == NULL)
+		return;
+	g_string_append_printf(str, "  %s: ", key);
+	for (gsize i = strlen(key); i < 20; i++)
+		g_string_append(str, " ");
+	g_string_append_printf(str, "%s\n", value);
+}
+static void
```

Comment:
newline needed

Reply:
Ah the missing TODO from earlier.  Those are identical for security and BIOS attrs.  need to actually move them out.